### ISSUE TITLE
Automatic complex type mapping

### ISSUE THREAD

**Issue by samardzicneo** [microsoft/node-api-dotnet#229]
I believe it would be quite useful for Node-in-.NET applications to support automatic mapping of .NET types to JS classes, like it should already be supported with JSMarshaller and TypeExporter for .NET-in-Node use cases.

**Comment by jasongin** [comment]
@samardzicneo Can you explain in more detail, and/or give an example of what you are proposing?

There is already fairly thorough automatic type mapping - what is the scenario that it does not work?
https://github.com/microsoft/node-api-dotnet?tab=readme-ov-file#automatic-efficient-marshaling

**Comment by samardzicneo** [comment]
@jasongin Yeah so basically I'm looking for something like the JSExport attribute as described there but in API form: Let me call an imaginary method `TypeMarshaller.RegisterType(Type type)` at runtime and it'd use the reflection code already part of the project (for .NET-in-Node) to map that type and any types it depends on into JS objects.

I've been having trouble with source generators in general and especially with support for dynamically loaded assemblies in one of the projects I'm trying to incorporate this in, source generators wouldn't even work.

**Comment by jasongin** [comment]
If I understand correctly, what you want is available via the following methods in the `JSMarshaller` class:

```C#
    /// <summary>
    /// Converts from a JS value to a specified type.
    /// </summary>
    /// <typeparam name="T">The type the value will be converted to.</typeparam>
    public T FromJS<T>(JSValue value);

    /// <summary>
    /// Converts from a specified type to a JS value.
    /// </summary>
    /// <typeparam name="T">The type the value will be converted from.</typeparam>
    public JSValue ToJS<T>(T value);

```

These methods generate code as necessary and compile a delegate (if not already in the cache) and then invoke it to convert between .NET and JavaScript types. The conversions are exactly the same as what you get from source generation, since both build-time and runtime code-generation are built on `JSMarshaller` which generates expression trees.

You can also access the same capability from the JavaScript side with the `dotnet.load(assembly)` method, which loads all of the assemblies namespaces and types onto the `dotnet` object, with automatic conversions between .NET and JS types. Though with I recommend using the type-definitions generation through MSBuild. See [dynamic-invoke.md](https://github.com/microsoft/node-api-dotnet/blob/main/Docs/dynamic-invoke.md)

Some of this is not documented very well and since there are many variations of .NET/JS interop scenarios I understand it can be unclear what is the best way to approach any task using this project. I'm working on writing some better documentation, but it's not ready yet.

**Comment by samardzicneo** [comment]
@jasongin With JSMarshaller specifically I got the impression that it's meant for .NET-in-Node and not for usage outside of that, since I've been receiving errors when using it related to dependency resolution. I don't have my laptop on hand for an exact exception but it was related to resolving the property types inside the class being mapped.

**Comment by jasongin** [comment]
> I got the impression that it's meant for .NET-in-Node and not for usage outside of that

`JSMarshaller` is used for both, because in either case objects have to be marshalled in both directions.

> I've been receiving errors when using it related to dependency resolution... it was related to resolving the property types inside the class being mapped

That could be a bug. We are still finding occasional issues in that area with more complex object graphs. Please share the error message and stack trace when you get a chance.

**Comment by samardzicneo** [comment]
@jasongin Sorry that I took a while, here's the error:

```
Microsoft.JavaScript.NodeApi.JSException: Type proxy not found for dependency: System.Int32
   at Microsoft.JavaScript.NodeApi.DotNetHost.TypeExporter.ExportTypeIfSupported(Type dependencyType, Boolean deferMembers) in /Users/samardzicneo/RiderProjects/nvite/Microsoft.JavaScript.NodeApi/DotNetHost/TypeExporter.cs:line 541
   at Microsoft.JavaScript.NodeApi.DotNetHost.TypeExporter.<>c__DisplayClass20_0.<ExportProperties>b__0(JSCallbackArgs args) in /Users/samardzicneo/RiderProjects/nvite/Microsoft.JavaScript.NodeApi/DotNetHost/TypeExporter.cs:line 621
   at Microsoft.JavaScript.NodeApi.JSValue.InvokeCallback[TDescriptor](napi_env env, napi_callback_info callbackInfo, JSValueScopeType scopeType, Func`2 getCallbackDescriptor) in /Users/samardzicneo/RiderProjects/nvite/Microsoft.JavaScript.NodeApi/JSValue.cs:line 1279
Error
   at Microsoft.JavaScript.NodeApi.JSValue.GetProperty(JSValue key) in /Users/samardzicneo/RiderProjects/nvite/Microsoft.JavaScript.NodeApi/JSValue.cs:line 487
   at Program.<>c.<<<Main>$>b__0_0>d.MoveNext() in /Users/samardzicneo/RiderProjects/nvite/nvite/Program.cs:line 39
--- End of stack trace from previous location ---
   at nvite.Library.Helpers.LibraryInterop.<>c__DisplayClass9_0.<<PostAsync>b__0>d.MoveNext() in /Users/samardzicneo/RiderProjects/nvite/nvite.Library/Helpers/LibraryInterop.cs:line 136
```

That is with the following code: 
```csharp
public class Test
{
    public int One { get; set; }
}

...

var te = new TypeExporter();
te.TryExportType(typeof(Test));

var obj = new Test();
var objJs = JSMarshaller.Current.ToJS(obj);

Console.WriteLine(objJs.GetProperty(JSValue.CreateStringUtf16("One")));
```

Note, I had to modify the TypeExporter class to be public instead of internal. Only using `JSMarshaller.ToJS` resulted in `Class not registered for JS export` errors.

**Comment by jasongin** [comment]
Thanks for the clarification. I understand now what you are trying to do, and how the current APIs do not support this at the moment. So far most design and testing of the automatic type mapping has been to support initiating it from the JS side. From this I can see that some improvements are necessary to make it usable from the .NET side as well.

I'll have to do some investigation to figure out the best way to accomplish this. Ideally I think the `TypeExporter` should remain an internal implementation detail, and `ToJS()` or perhaps a higher-level method should automatically export any types as necessary.

In case you're wondering: What does "exporting" a .NET class do?
 - An equivalent namespaced JS class prototype is defined, with a constructor function that calls back to the .NET constructor.
 - Static and instance properties and methods are defined on the class prototype. Initially all of them are stubs, but on first access each property gets redefined to call back to the corresponding .NET property or method.
 - The mapping between .NET type and JS class/constructor object is registered with the `JSRuntimeContext`, for use in any marshalling operations.
 
> Type proxy not found for dependency: System.Int32

For now, you can resolve this with the following code:
```C#
te.ExportAssemblyTypes(typeof(object).Assembly, new JSObject());
```

**Comment by jasongin** [comment]
Fixed in version 0.6.

You can use code similar to [this test case](https://github.com/microsoft/node-api-dotnet/blob/5cce1cc202a5c3f0463e4e05076d89bde6f8c836/test/NodejsEmbeddingTests.cs#L237).
